larpandoracontent/LArObjects/LArDiscreteProbabilityVector.cc

```diff
@@ -0,0 +1,237 @@
+    float cumulativeProbability(0.f);
+    for (size_t iElement = 0; iElement < discreteProbabilityVector.GetSize(); ++iElement)
+    {
+        const size_t randomElementIndex(randomisedElements.at(iElement));
+        const float deltaX(discreteProbabilityVector.GetWidth(randomElementIndex));
+        const float probabilityDensity(discreteProbabilityVector.GetProbabilityDensity(randomElementIndex));
+        cumulativeProbability+=probabilityDensity*(m_useWidths ? deltaX : 1.f);
+        randomisedProbabilityData.emplace_back(DiscreteProbabilityVector::DiscreteProbabilityDatum(xPos, probabilityDensity, 
+            cumulativeProbability, deltaX));
+        xPos+=deltaX;
+    }
+
+    return randomisedProbabilityData;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+template <typename TX, typename TY>
+bool DiscreteProbabilityVector::SortInputDataByX(const InputDatum<TX, TY> &lhs, const InputDatum<TX, TY> &rhs)
+{
+    const float deltaX(static_cast<float>(rhs.first) - static_cast<float>(lhs.first));
+    if (std::fabs(deltaX) < std::numeric_limits<float>::epsilon())
+        return (lhs.second < rhs.second);
+
+    return (lhs.first < rhs.first);
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+template <typename TX, typename TY>
+float DiscreteProbabilityVector::CalculateNormalisation(const InputData<TX, TY> &inputData) const
+{
+    if (2 > inputData.size())
+        throw pandora::StatusCodeException(pandora::STATUS_CODE_INVALID_PARAMETER);
+
+    float normalisation(0.f);
+
+    for (size_t iDatum = 0; iDatum < inputData.size()-1; ++iDatum)
+    {
+        const float y(static_cast<float>(inputData.at(iDatum).second));
+        normalisation += y*(m_useWidths ? 
+            static_cast<float>(inputData.at(iDatum+1).first) - static_cast<float>(inputData.at(iDatum).first) : 1.f);
+    }
+    const float y(static_cast<float>(inputData.back().second));
+    normalisation += y*(m_useWidths ? m_xUpperBound - static_cast<float>(inputData.back().first) : 1.f);
+
+    return normalisation;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+template DiscreteProbabilityVector::DiscreteProbabilityVector(const InputData<int, float> &, int const, bool const);
+template DiscreteProbabilityVector::DiscreteProbabilityVector(const InputData<float, int> &, float const, bool const);
+template DiscreteProbabilityVector::DiscreteProbabilityVector(const InputData<float, float> &, float const, bool const);
+template DiscreteProbabilityVector::DiscreteProbabilityVector(const InputData<int, int> &, int const, bool const);
+
+template DiscreteProbabilityVector::DiscreteProbabilityData DiscreteProbabilityVector::InitialiseDiscreteProbabilityData(
```

Comment:
Personally prefer to keep these each on one line.

Reply:
Done